Source/Core/DolphinQt/DiscordHandler.cpp

```diff
@@ -0,0 +1,84 @@
+// Copyright 2018 Dolphin Emulator Project
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
+#ifdef USE_DISCORD_PRESENCE
+
+#include "DolphinQt/DiscordHandler.h"
+
+#include <iterator>
+
+#include <QApplication>
+
+#include "Common/Thread.h"
+
+#include "UICommon/DiscordPresence.h"
+
+#include "DolphinQt/DiscordJoinRequestDialog.h"
+
+DiscordHandler::DiscordHandler(QWidget* parent) : QObject{parent}, m_parent{parent}
+{
+  connect(this, &DiscordHandler::JoinRequest, this, &DiscordHandler::ShowNewJoinRequest);
+}
+
+DiscordHandler::~DiscordHandler()
+{
+  Stop();
+}
+
+void DiscordHandler::Start()
+{
+  m_stop_requested.Set(false);
+  m_thread = std::thread(&DiscordHandler::Run, this);
+}
+
+void DiscordHandler::Stop()
+{
+  m_stop_requested.Set(true);
+
+  if (m_thread.joinable())
+    m_thread.join();
+}
+
+void DiscordHandler::DiscordJoinRequest(const char* id, const std::string& discord_tag,
+                                        const char* avatar)
+{
+  emit DiscordHandler::JoinRequest(id, discord_tag, avatar);
+}
+
+void DiscordHandler::DiscordJoin()
+{
+  emit DiscordHandler::Join();
+}
+
+void DiscordHandler::ShowNewJoinRequest(const std::string& id, const std::string& discord_tag,
+                                        const std::string& avatar)
+{
+  m_request_dialogs.emplace_front(m_parent, id, discord_tag, avatar);
+  m_request_dialogs.front().show();
```

Comment:
We should also call `raise()` and `activateWindow()` on the dialog, otherwise it tends to just open behind other windows and goes unnoticed.